diff -Naur openmp-14.0.6.src/libomptarget/plugins/common/elf_common/CMakeLists.txt patched/openmp-14.0.6.src/libomptarget/plugins/common/elf_common/CMakeLists.txt
--- openmp-14.0.6.src/libomptarget/plugins/common/elf_common/CMakeLists.txt	2022-06-22 09:46:24.000000000 -0700
+++ patched/openmp-14.0.6.src/libomptarget/plugins/common/elf_common/CMakeLists.txt	2025-08-24 03:30:01.678093824 -0700
@@ -16,9 +16,6 @@
 set_property(TARGET elf_common PROPERTY POSITION_INDEPENDENT_CODE ON)
 llvm_update_compile_flags(elf_common)
 set(LINK_LLVM_LIBS LLVMBinaryFormat LLVMObject LLVMSupport)
-if (LLVM_LINK_LLVM_DYLIB)
-  set(LINK_LLVM_LIBS LLVM)
-endif()
 target_link_libraries(elf_common INTERFACE ${LINK_LLVM_LIBS})
 include_directories(${LIBOMPTARGET_LLVM_INCLUDE_DIRS})
 add_dependencies(elf_common ${LINK_LLVM_LIBS})
