From 9086199f307140fed7dc97ddce63e36a2de6da3a Mon Sep 17 00:00:00 2001
From: "Todd A. Anderson" <drtodd13@comcast.net>
Date: Tue, 22 Aug 2023 16:01:52 -0700
Subject: [PATCH] Fix crosscompiling LLVM tools

---
 clang/CMakeLists.txt | 10 ++++++++++
 1 file changed, 10 insertions(+)

diff --git a/clang/CMakeLists.txt b/clang/CMakeLists.txt
index 937a8467df1d..79e310c636b0 100644
--- a/clang/CMakeLists.txt
+++ b/clang/CMakeLists.txt
@@ -114,6 +114,16 @@ if(CLANG_BUILT_STANDALONE)
   option(LLVM_ENABLE_LIBXML2 "Use libxml2 if available." ON)
 
   include(AddLLVM)
+  if(CMAKE_CROSSCOMPILING)
+    set(LLVM_USE_HOST_TOOLS ON)
+    include(CrossCompile)
+    if (NOT NATIVE_LLVM_DIR)
+      message(FATAL_ERROR
+       "Crosscompiling standalone requires the variable NATIVE_LLVM_DIR
+       for building the native lldb-tblgen used during the build process.")
+    endif()
+    llvm_create_cross_target(Clang NATIVE "" Release -DLLVM_DIR=${NATIVE_LLVM_DIR})
+  endif()
   include(TableGen)
   include(HandleLLVMOptions)
   include(VersionFromVCS)
-- 
2.31.1

